### PR TITLE
Improve and merge `Embedded` and `Nested` 

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | -

### CODE REVIEW

File: src/Rule/NestedHandler.php [hunk 1]
```diff
@@ -58,26 +59,44 @@ public function validate(mixed $value, object $rule, ValidationContext $context)
...
         }
 
-        $compoundResult = new Result();
-        if (!is_object($value) && !is_array($value)) {
-            $message = sprintf('Value should be an array or an object. %s given.', gettype($value));
+        if ($rule->getRules() === null) {
+            if (!is_object($value)) {
+                throw new InvalidArgumentException(
+                    sprintf(
+                        'Nested rule without rules available for objects only. %s given.',
```
Comment (samdark) [inline src/Rule/NestedHandler.php:66]:
```suggestion
                        'Nested rule without rules could be used for objects only. %s given.',
```

**Comment by samdark** [comment]
The rest seems to be 👍

File: src/Rule/Nested.php [hunk 2]
```diff
@@ -38,52 +41,65 @@ final class Nested implements SerializableRuleInterface, BeforeValidationInterfa
...
+         * @var int
+         */
+        private int $propertyVisibility = ReflectionProperty::IS_PRIVATE | ReflectionProperty::IS_PROTECTED | ReflectionProperty::IS_PUBLIC,
+
+        /**
+         * This options use on parse rules from class defined in {@see $rules}.
+         *
+         * @var int
+         */
+        private int $rulesPropertyVisibility = ReflectionProperty::IS_PRIVATE | ReflectionProperty::IS_PROTECTED | ReflectionProperty::IS_PUBLIC,
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/Rule/Nested.php:69]: This is should change separately, it's properties for different things..

1) `propertyVisibility` use for object value for get data and rules from object.

2) `rulesPropertyVisibility` use for get rules from class that setted in `rules` property.

May be there are throughts on how to rename properties more clearly?

File: src/Rule/Nested.php [hunk 3]
```diff
@@ -38,52 +41,65 @@ final class Nested implements SerializableRuleInterface, BeforeValidationInterfa
...
 
+    /**
+     * @var iterable<Closure|Closure[]|RuleInterface|RuleInterface[]>|null
+     */
+    private ?iterable $rules = null;
+
     public function __construct(
         /**
-         * @var iterable<\Closure|\Closure[]|RuleInterface|RuleInterface[]>
+         * Null available only for objects.
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/Rule/Nested.php:51]: Agree. I'm add more detailed comment.

[the remark this answers is not in the document]
Comment (vjik) [reply src/Rule/Nested.php]: Done

File: src/Rule/Nested.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (samdark) [reply src/Rule/Nested.php:76]: Some extra phpdoc would do. But the question number 3 remains: when do we need to change these visibility options?

[the remark this answers is not in the document]
Comment (vjik) [reply src/Rule/Nested.php:76]: > Some extra phpdoc would do. But the question number 3 remains: when do we need to change these visibility options?

For example, we need get data from public properties only and don't parse private and protected properties of validated object.

[the remark this answers is not in the document]
Comment (samdark) [reply src/Rule/Nested.php:76]: OK. Does it make sense to read rules from private properties but not read data from the same properties?

[the remark this answers is not in the document]
Comment (vjik) [reply src/Rule/Nested.php:76]: > OK. Does it make sense to read rules from private properties but not read data from the same properties?

This options for different classes. For example:

```
#[Nested(rules:ClassWithRules::name)
private AnotherObject $object;
```

`rulesPropertyVisibility` used for `ClassWithRules`.
`propertyVisibility` used for `AnotherObject`.

**Review by samdark — APPROVED** [review]

**Pull request merged**